src/connectionController.ts

```diff
@@ -77,106 +98,132 @@ export default class ConnectionController {
   }
 
-  async loadSavedConnections(): Promise<void> {
-    if (!ext.keytarModule) {
+  private async _loadSavedConnectionsByStore(
+    savedConnections: ConnectionsFromStorage
+  ): Promise<void> {
+    if (!savedConnections || !Object.keys(savedConnections).length) {
       return;
     }
 
-    // Load saved connections from storage.
-    const existingGlobalConnections: SavedConnection[] =
-      this._storageController.get(StorageVariables.GLOBAL_SAVED_CONNECTIONS) ||
-      {};
+    /** User connections are being saved both in:
+     * 1. Vscode global/workspace storage (without secrets) + keychain (secrets)
+     * 2. Memory of the extension (with secrets)
+     */
+    await Promise.all(
+      Object.keys(savedConnections).map(async (connectionId) => {
+        // Get connection info from vscode storage and merge with secrets.
+        const connectionInfoWithSecrets = await this._getConnectionInfoWithSecrets(
+          savedConnections[connectionId]
+        );
+
+        // Save connection info with secrets to extension memory.
+        if (connectionInfoWithSecrets) {
+          this._connections[connectionId] = connectionInfoWithSecrets;
+        }
 
-    if (Object.keys(existingGlobalConnections).length > 0) {
-      // Try to pull in the connections previously saved globally on vscode.
-      await Promise.all(
-        Object.keys(existingGlobalConnections).map((connectionId) =>
-          this._loadSavedConnection(
-            connectionId,
-            existingGlobalConnections[connectionId]
-          )
-        )
-      );
-    }
+        this.eventEmitter.emit(DataServiceEventTypes.CONNECTIONS_DID_CHANGE);
+      })
+    );
+  }
 
-    const existingWorkspaceConnections: SavedConnection[] =
-      this._storageController.get(
-        StorageVariables.WORKSPACE_SAVED_CONNECTIONS,
-        StorageScope.WORKSPACE
-      ) ||
-      {};
-
-    if (Object.keys(existingWorkspaceConnections).length > 0) {
-      // Try to pull in the connections previously saved on the workspace.
-      await Promise.all(
-        Object.keys(existingWorkspaceConnections).map((connectionId) =>
-          this._loadSavedConnection(
-            connectionId,
-            existingWorkspaceConnections[connectionId]
-          )
-        )
-      );
-    }
+  async loadSavedConnections(): Promise<void> {
+    // Try to pull in the connections previously saved in the global storage of vscode.
+    const existingGlobalConnections = this._storageController.get(
+      StorageVariables.GLOBAL_SAVED_CONNECTIONS,
+      StorageLocation.GLOBAL
+    );
+    await this._loadSavedConnectionsByStore(existingGlobalConnections);
+
+    // Try to pull in the connections previously saved in the workspace storage of vscode.
+    const existingWorkspaceConnections = this._storageController.get(
+      StorageVariables.WORKSPACE_SAVED_CONNECTIONS,
+      StorageLocation.WORKSPACE
+    );
+    await this._loadSavedConnectionsByStore(existingWorkspaceConnections);
```

Comment:
(nit:) Since we’re already parallelizing this above, could/should these also be done in parallel?

```suggestion
    await Promise.all([(async() => {
      // Try to pull in the connections previously saved in the global storage of vscode.
      const existingGlobalConnections = this._storageController.get(
        StorageVariables.GLOBAL_SAVED_CONNECTIONS,
        StorageLocation.GLOBAL
      );
      await this._loadSavedConnectionsByStore(existingGlobalConnections);
    })(), (async() => {
      // Try to pull in the connections previously saved in the workspace storage of vscode.
      const existingWorkspaceConnections = this._storageController.get(
        StorageVariables.WORKSPACE_SAVED_CONNECTIONS,
        StorageLocation.WORKSPACE
      );
      await this._loadSavedConnectionsByStore(existingWorkspaceConnections);
    })()]);
```